src/embed_tests/TestDomainReload.cs

```diff
@@ -234,6 +453,162 @@ static Assembly ResolveAssembly(object sender, ResolveEventArgs args)
+    //    using GIL { Python.Exec(pyScript); }
+    // but the rest is useful for debugging.
+    //
+    // What matters in the python code is gc.collect and clr.AddReference.
+    //
+    // Note that the language version is 2.0, so no $"foo{bar}" syntax.
+    //
+    static class PythonRunner
+    {
+        public static void RunPython()
+        {
+            AppDomain.CurrentDomain.DomainUnload += OnDomainUnload;
+            string name = AppDomain.CurrentDomain.FriendlyName;
+            Console.WriteLine(string.Format("[{0} in .NET] In PythonRunner.RunPython", name));
+            var mode = PythonEngine.DefaultShutdownMode;
+            if (mode == ShutdownMode.Normal)
+            {
+                mode = ShutdownMode.Soft;
+            }
+            PythonEngine.Initialize(mode: mode);
+            try
+            {
+                using (Py.GIL())
+                {
+                    try
+                    {
+                        var pyScript = string.Format("import clr\n"
+                            + "print('[{0} in python] imported clr')\n"
+                            + "clr.AddReference('System')\n"
+                            + "print('[{0} in python] allocated a clr object')\n"
+                            + "import gc\n"
+                            + "gc.collect()\n"
+                            + "print('[{0} in python] collected garbage')\n",
+                            name);
+                        PythonEngine.Exec(pyScript);
+                    }
+                    catch (Exception e)
+                    {
+                        Console.WriteLine(string.Format("[{0} in .NET] Caught exception: {1}", name, e));
+                        throw;
+                    }
+                }
+            }
+            finally
+            {
+                PythonEngine.BeginAllowThreads();
+            }
+        }
+
+
+        private static IntPtr _state;
+
+        public static void InitPython(ShutdownMode mode)
+        {
+            PythonEngine.Initialize(mode: mode);
+            _state = PythonEngine.BeginAllowThreads();
+        }
+
+        public static void ShutdownPython()
+        {
+            PythonEngine.EndAllowThreads(_state);
+            PythonEngine.Shutdown();
+        }
+
+        public static void ShutdownPythonCompletely()
+        {
+            PythonEngine.EndAllowThreads(_state);
+            // XXX: Reload mode will reserve clr objects after `Runtime.Shutdown`,
+            // if it used a another mode(the default mode) in other tests,
+            // when other tests trying to access these reserved objects, it may cause Domain exception,
+            // thus it needs to reduct to Soft mode to make sure all clr objects remove from Python.
+            var defaultMode = PythonEngine.DefaultShutdownMode;
+            if (defaultMode != ShutdownMode.Reload)
+            {
+                PythonEngine.ShutdownMode = defaultMode;
+            }
```

Comment:
I added the `ShutdownMode` argument to `Shutdown'`. Because it's not possible (for example) to go from `Normal` to `Reload`, we have to remember the mode we were initialized in. See https://github.com/pythonnet/pythonnet/pull/958/commits/25a4064752a2708322dc9867caa7531b0a93ef60 for full details.